proxy/http/HttpTransact.cc

```diff
@@ -1546,6 +1546,36 @@ HttpTransact::OSDNSLookup(State *s)
     TRANSACT_RETURN(SM_ACTION_SEND_ERROR_CACHE_NOOP, nullptr);
   }
 
+  if (s->redirect_info.redirect_in_process) {
+    RedirectEnabled::Action action = RedirectEnabled::Action::INVALID;
+    if (true == Machine::instance()->is_self(s->host_db_info.ip())) {
+      action = s->http_config_param->redirect_actions_self_action;
+    } else {
+      ink_release_assert(s->http_config_param->redirect_actions_map != nullptr);
+      ink_release_assert(
+        s->http_config_param->redirect_actions_map->contains(s->host_db_info.ip(), reinterpret_cast<void **>(&action)));
+    }
+    switch (action) {
+    case RedirectEnabled::Action::FOLLOW:
+      TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect address. Following");
+      break;
+    case RedirectEnabled::Action::REJECT:
+      TxnDebug("http_trans", "[OSDNSLookup] Invalid redirect address. Rejecting.");
+      build_error_response(s, HTTP_STATUS_FORBIDDEN, nullptr, "request#syntax_error");
+      SET_VIA_STRING(VIA_DETAIL_TUNNEL, VIA_DETAIL_TUNNEL_NO_FORWARD);
+      TRANSACT_RETURN(SM_ACTION_SEND_ERROR_CACHE_NOOP, nullptr);
+      break;
+    case RedirectEnabled::Action::RETURN:
```

Comment:
Why not disable redirect for the transaction and return the origins redirect response?

Reply:
I think this is because we are already in the process of redirecting, and so disabling redirect for the transaction here would be too late.

This is tricky because of the way the state machine transitions are set up. What we need to do is resolve the target (`Location`) of the redirect response and then check the resulting address against our configuration in order to figure out what the administrator wants to do. But we only proceed to the DNS lookup state *after* we have "committed" to following the redirect. 

* If we disable redirect for the transaction too early, we will not be able to resolve the host and we will potentially return responses we should have followed or rejected. 
* If we wait until after host resolution to disable redirect for the transaction, then the logic will have no effect, and we will potentially follow redirects that we should have returned or rejected.

Reply:
@d2r If you want the process to die on a error I would use Fatal.  It is used all over traffic_server for logging Fatal errors and exiting.

Reply:
> @d2r If you want the process to die on a error I would use Fatal. It is used all over traffic server for logging Fatal errors and exiting.

Created #4229 as discussed so that there is clarity on why `Emergency` exists apart from `Fatal`.